File: frontend/src/types/skill.ts
/**
 * 全局 TypeScript 类型
 */

/** Sprint S02: 来源类型枚举 */
export type SourceType = 'LOCAL_ZIP' | 'LOCAL_FILE' | 'GIT_URL' | null

/** Sprint S02: 同步状态 */
export type SyncStatus = 'success' | 'failed' | 'syncing' | null

export interface Skill {
  id: number
  name: string
  slug: string
  displayName?: string
  description: string
  body?: string
  categoryId?: number
  categoryName?: string
  categorySlug?: string
  // S18: USAGE 维度（保留平铺字段向后兼容）
  usageCategoryId?: number
  usageCategoryName?: string
  usageCategorySlug?: string
  // S24: USAGE 嵌套节点（含父类目，前端 chip 配色按 parentCode 取色）
  usageCategory?: UsageCategoryNode
  tags?: string[]
  license?: string
  allowedTools?: string
  compatibility?: string
  version?: string
  homepage?: string
  authorName?: string
  authorEmail?: string
  authorGithub?: string
  icon?: string
  source?: string
  installCommand?: string
  downloadUrl?: string
  packageSize?: number
  stars?: number
  installs?: number
  views?: number
  ratingAvg?: number
  ratingCount?: number
  status?: string
  featured?: boolean
  favorited?: boolean
  createTime?: string
  updateTime?: string
  // ===== Sprint S02: Git 源字段 =====
  sourceType?: SourceType
  sourceUrl?: string
  sourceRef?: string
  tokenHint?: string           // 前端展示用脱敏
  lastSyncAt?: string
  lastSyncStatus?: SyncStatus
  lastSyncError?: string
  lastCommitSha?: string
}

export interface Category {
  id: number
  name: string
  slug: string
  description?: string
  icon?: string
  sortOrder?: number
  skillCount?: number
  // S18: 分类维度（SOC=职业 / USAGE=用途）
  type?: 'SOC' | 'USAGE'
  code?: string
  parentId?: number
  children?: Category[]
}

/** S24: USAGE 嵌套节点（前端 chip 用，按 parentCode 取色） */
export interface UsageCategoryNode {
  id: number
  code: string           // 二级 code，如 PURPOSE-DEV-FRONTEND
  name: string           // 二级中文名，如 "前端开发"
  slug?: string
  description?: string   // 父类目 description
  parentId?: number      // 父类目 id（一级 USAGE 的 id）
  parentCode?: string    // 父类目 code，如 PURPOSE-DEV（配色 key）
  parentName?: string    // 父类目中文名，如 "开发"
}

export interface Tag {
  id: number
  name: string
  slug: string
  skillCount?: number
}

export interface Review {
  id: number
  skillId: number
  userId: number
  username?: string
  userAvatar?: string
  rating: number
  comment?: string
  createTime?: string
}

export interface PageResult<T> {
  records: T[]
  total: number
  page: number
  size: number
}

// ===== Sprint S02: Git 源相关类型 =====

/** POST /api/admin/skills/from-git 请求体 */
export interface GitImportRequest {
  url: string
  ref?: string
  username?: string
  token?: string
  insecureSkipTls?: boolean
}

/** POST /api/admin/skills/from-git 响应 */
export interface GitImportResult {
  repoUrl: string
  ref?: string
  workdir?: string
  totalDiscovered: number
  totalImported: number
  totalSkipped: number
  skipReasons?: string[]
  discovered: Array<{
    name: string
    path: string
    description?: string
    version?: string
    action: 'created' | 'updated' | 'skipped'
    skipReason?: string
    skillId?: number
  }>
}

/** POST /api/admin/skills/{id}/sync 响应 */
export interface GitSyncResult {
  id: number
  lastCommitSha?: string
  changed: boolean
  message: string
}

/** GET /api/admin/skills/{id}/sync-status 响应 */
export interface GitSyncStatus {
  id: number
  name: string
  sourceUrl: string
  sourceRef?: string
  tokenHint?: string
  lastSyncAt?: string
  lastSyncStatus?: SyncStatus
  lastSyncError?: string
  lastCommitSha?: string
}

File: backend/src/main/java/com/meiya/skillsmap/entity/Skill.java
package com.meiya.skillsmap.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableField;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableLogic;
import com.baomidou.mybatisplus.annotation.TableName;

import java.io.Serializable;
import java.time.LocalDateTime;

@TableName("skill")
public class Skill implements Serializable {

    @TableId(type = IdType.AUTO)
    private Long id;

    @TableField("name")
    private String name;

    @TableField("slug")
    private String slug;

    @TableField("display_name")
    private String displayName;

    @TableField("description")
    private String description;

    @TableField("body")
    private String body;

    @TableField("category_id")
    private Long categoryId;

    // S18: USAGE 维度 FK（nullable，可只挂 SOC 或只挂 USAGE 或都不挂）
    @TableField("usage_category_id")
    private Long usageCategoryId;

    @TableField("tags")
    private String tags;

    @TableField("license")
    private String license;

    @TableField("allowed_tools")
    private String allowedTools;

    @TableField("compatibility")
    private String compatibility;

    @TableField("metadata")
    private String metadata;

    @TableField("version")
    private String version;

    @TableField("homepage")
    private String homepage;

    @TableField("author_name")
    private String authorName;

    @TableField("author_email")
    private String authorEmail;

    @TableField("author_github")
    private String authorGithub;

    @TableField("icon")
    private String icon;

    @TableField("source")
    private String source;

    @TableField("install_command")
    private String installCommand;

    @TableField("download_url")
    private String downloadUrl;

    @TableField("package_size")
    private Long packageSize;

    @TableField("stars")
    private Integer stars;

    @TableField("installs")
    private Integer installs;

    @TableField("views")
    private Integer views;

    @TableField("rating_avg")
    private Double ratingAvg;

    @TableField("rating_count")
    private Integer ratingCount;

    @TableField("status")
    private String status;

    @TableField("featured")
    private Boolean featured;

    // ===== Sprint S02: Git 源字段 =====
    /** 来源类型: LOCAL_ZIP / LOCAL_FILE / GIT_URL / null (历史数据视作 LOCAL_ZIP) */
    @TableField("source_type")
    private String sourceType;

    /** Git 仓库 URL（仅 GIT_URL 用） */
    @TableField("source_url")
    private String sourceUrl;

    /** Branch / Tag / Commit SHA（仅 GIT_URL 用） */
    @TableField("source_ref")
    private String sourceRef;

    /** Jasypt 加密后的 username:token（仅 GIT_URL 用，私有仓库）；前端 @JsonIgnore */
    @TableField("source_token_enc")
    private String sourceTokenEnc;

    /** Token 脱敏提示，如 ghp_xx****abcd（仅展示用） */
    @TableField("token_hint")
    private String tokenHint;

    /** 最近一次成功同步时间（手动 / 定时） */
    @TableField("last_sync_at")
    private LocalDateTime lastSyncAt;

    /** 最近一次同步状态: success / failed / syncing */
    @TableField("last_sync_status")
    private String lastSyncStatus;

    /** 最近一次同步错误信息（失败时） */
    @TableField("last_sync_error")
    private String lastSyncError;

    /** 最近一次同步后的 HEAD commit SHA */
    @TableField("last_commit_sha")
    private String lastCommitSha;

    @TableField("created_by_user_id")
    private Long createdByUserId;

    // ===== Sprint S38: 用户上传端点新增字段 =====
    /** 上传者 user id（FK → user.id ON DELETE SET NULL）。
     *  与 created_by_user_id 解耦：MVP 暂存同值；ADMIN 代上传时二者不同。 */
    @TableField("uploader_user_id")
    private Long uploaderUserId;

    @TableField("create_time")
    private LocalDateTime createTime;

    @TableField("update_time")
    private LocalDateTime updateTime;

    @TableLogic
    @TableField("deleted")
    private Integer deleted;

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public String getSlug() { return slug; }
    public void setSlug(String slug) { this.slug = slug; }
    public String getDisplayName() { return displayName; }
    public void setDisplayName(String displayName) { this.displayName = displayName; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public String getBody() { return body; }
    public void setBody(String body) { this.body = body; }
    public Long getCategoryId() { return categoryId; }
    public void setCategoryId(Long categoryId) { this.categoryId = categoryId; }
    public Long getUsageCategoryId() { return usageCategoryId; }
    public void setUsageCategoryId(Long usageCategoryId) { this.usageCategoryId = usageCategoryId; }
    public String getTags() { return tags; }
    public void setTags(String tags) { this.tags = tags; }
    public String getLicense() { return license; }
    public void setLicense(String license) { this.license = license; }
    public String getAllowedTools() { return allowedTools; }
    public void setAllowedTools(String allowedTools) { this.allowedTools = allowedTools; }
    public String getCompatibility() { return compatibility; }
    public void setCompatibility(String compatibility) { this.compatibility = compatibility; }
    public String getMetadata() { return metadata; }
    public void setMetadata(String metadata) { this.metadata = metadata; }
    public String getVersion() { return version; }
    public void setVersion(String version) { this.version = version; }
    public String getHomepage() { return homepage; }
    public void setHomepage(String homepage) { this.homepage = homepage; }
    public String getAuthorName() { return authorName; }
    public void setAuthorName(String authorName) { this.authorName = authorName; }
    public String getAuthorEmail() { return authorEmail; }
    public void setAuthorEmail(String authorEmail) { this.authorEmail = authorEmail; }
    public String getAuthorGithub() { return authorGithub; }
    public void setAuthorGithub(String authorGithub) { this.authorGithub = authorGithub; }
    public String getIcon() { return icon; }
    public void setIcon(String icon) { this.icon = icon; }
    public String getSource() { return source; }
    public void setSource(String source) { this.source = source; }
    public String getInstallCommand() { return installCommand; }
    public void setInstallCommand(String installCommand) { this.installCommand = installCommand; }
    public String getDownloadUrl() { return downloadUrl; }
    public void setDownloadUrl(String downloadUrl) { this.downloadUrl = downloadUrl; }
    public Long getPackageSize() { return packageSize; }
    public void setPackageSize(Long packageSize) { this.packageSize = packageSize; }
    public Integer getStars() { return stars; }
    public void setStars(Integer stars) { this.stars = stars; }
    public Integer getInstalls() { return installs; }
    public void setInstalls(Integer installs) { this.installs = installs; }
    public Integer getViews() { return views; }
    public void setViews(Integer views) { this.views = views; }
    public Double getRatingAvg() { return ratingAvg; }
    public void setRatingAvg(Double ratingAvg) { this.ratingAvg = ratingAvg; }
    public Integer getRatingCount() { return ratingCount; }
    public void setRatingCount(Integer ratingCount) { this.ratingCount = ratingCount; }
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    public Boolean getFeatured() { return featured; }
    public void setFeatured(Boolean featured) { this.featured = featured; }
    public Long getCreatedByUserId() { return createdByUserId; }
    public void setCreatedByUserId(Long createdByUserId) { this.createdByUserId = createdByUserId; }
    public Long getUploaderUserId() { return uploaderUserId; }
    public void setUploaderUserId(Long uploaderUserId) { this.uploaderUserId = uploaderUserId; }
    public String getSourceType() { return sourceType; }
    public void setSourceType(String sourceType) { this.sourceType = sourceType; }
    public String getSourceUrl() { return sourceUrl; }
    public void setSourceUrl(String sourceUrl) { this.sourceUrl = sourceUrl; }
    public String getSourceRef() { return sourceRef; }
    public void setSourceRef(String sourceRef) { this.sourceRef = sourceRef; }
    public String getSourceTokenEnc() { return sourceTokenEnc; }
    public void setSourceTokenEnc(String sourceTokenEnc) { this.sourceTokenEnc = sourceTokenEnc; }
    public String getTokenHint() { return tokenHint; }
    public void setTokenHint(String tokenHint) { this.tokenHint = tokenHint; }
    public LocalDateTime getLastSyncAt() { return lastSyncAt; }
    public void setLastSyncAt(LocalDateTime lastSyncAt) { this.lastSyncAt = lastSyncAt; }
    public String getLastSyncStatus() { return lastSyncStatus; }
    public void setLastSyncStatus(String lastSyncStatus) { this.lastSyncStatus = lastSyncStatus; }
    public String getLastSyncError() { return lastSyncError; }
    public void setLastSyncError(String lastSyncError) { this.lastSyncError = lastSyncError; }
    public String getLastCommitSha() { return lastCommitSha; }
    public void setLastCommitSha(String lastCommitSha) { this.lastCommitSha = lastCommitSha; }
    public LocalDateTime getCreateTime() { return createTime; }
    public void setCreateTime(LocalDateTime createTime) { this.createTime = createTime; }
    public LocalDateTime getUpdateTime() { return updateTime; }
    public void setUpdateTime(LocalDateTime updateTime) { this.updateTime = updateTime; }
    public Integer getDeleted() { return deleted; }
    public void setDeleted(Integer deleted) { this.deleted = deleted; }
}
